ports/stm32/can.c

```diff
@@ -312,119 +310,124 @@ HAL_StatusTypeDef CAN_Transmit(CAN_HandleTypeDef *hcan, uint32_t Timeout) {
     }
 }
 
-static void can_rx_irq_handler(uint can_id, uint fifo_id) {
-    mp_obj_t callback;
-    pyb_can_obj_t *self;
-    mp_obj_t irq_reason = MP_OBJ_NEW_SMALL_INT(0);
-    byte *state;
-
-    self = MP_STATE_PORT(pyb_can_obj_all)[can_id - 1];
-
-    if (fifo_id == CAN_FIFO0) {
-        callback = self->rxcallback0;
-        state = &self->rx_state0;
+static void can_rx_irq_handler(uint can_id, CAN_TypeDef *instance, can_rx_fifo_t fifo) {
+    uint32_t full_flag, full_int, overrun_flag, overrun_int, pending_int;
+    __IO uint32_t *fifo_reg;
+
+    if (fifo == CAN_RX_FIFO0) {
+        fifo_reg = &instance->RF0R;
+        full_flag = CAN_RF0R_FULL0;
+        full_int = CAN_IER_FFIE0;
+        overrun_flag = CAN_RF0R_FOVR0;
+        overrun_int = CAN_IER_FOVIE0;
+        pending_int = CAN_IER_FMPIE0;
     } else {
-        callback = self->rxcallback1;
-        state = &self->rx_state1;
+        fifo_reg = &instance->RF1R;
+        full_flag = CAN_RF1R_FULL1;
+        full_int = CAN_IER_FFIE1;
+        overrun_flag = CAN_RF1R_FOVR1;
+        overrun_int = CAN_IER_FOVIE1;
+        pending_int = CAN_IER_FMPIE1;
     }
 
-    switch (*state) {
-        case RX_STATE_FIFO_EMPTY:
-            __HAL_CAN_DISABLE_IT(&self->can,  (fifo_id == CAN_FIFO0) ? CAN_IT_FMP0 : CAN_IT_FMP1);
-            irq_reason = MP_OBJ_NEW_SMALL_INT(0);
-            *state = RX_STATE_MESSAGE_PENDING;
-            break;
-        case RX_STATE_MESSAGE_PENDING:
-            __HAL_CAN_DISABLE_IT(&self->can, (fifo_id == CAN_FIFO0) ? CAN_IT_FF0 : CAN_IT_FF1);
-            __HAL_CAN_CLEAR_FLAG(&self->can, (fifo_id == CAN_FIFO0) ? CAN_FLAG_FF0 : CAN_FLAG_FF1);
```

Comment:
I can see where `__HAL_CAN_DISABLE_IT` has been refactored to (the `instance->IER &= ~...` code below).

But, I can't see where `__HAL_CAN_CLEAR_FLAG` has been moved to.  For FDCAN it's replaced with `instance->IR` but for standard CAN I don't see any place that clears the interrupt flags.

Reply:
It was intended to be done on the next line:

```c
        instance->IER &= ~full_int;
        *fifo_reg &= ~full_flag;
```

`fifo_reg` is either RF0R or RF1R and `full_flag` is either `CAN_RF0R_FULL0` or `CAN_RF1R_FULL1`. This is the equivalent to the register.

However this is wrong as this bit is "write one to clear", will fix and make the behaviour clearer.

Reply:
I've rewritten this to use the HAL macros again, as less prone to mistakes like missing "write 1 to clear". See what you think.

Reply:
Looks good now.